### ISSUE TITLE
Uninformative error message from dependency resolver.

### ISSUE THREAD

**Issue by NightBlues** [ocaml/opam#3912]
I can not reopen issue https://github.com/ocaml/opam/issues/3083 but its still here.
For 4.08 ocaml I do:
```
$ opam install -vvv ppx_deriving
+ /home/nightblues/.opam/4.08.0+flambda/bin/ocamlc "-vnum"
- 4.08.0
Sorry, no solution found: there seems to be a problem with your request.

No solution found, exiting
'opam install -vvv ppx_deriving' failed.
```
The reason for this case is that ppx_deriving depends on ppx_tools, that has ocaml < 4.08 dependency, but I've found it manually.
```
$ opam config report
# opam config report
# opam-version      2.0.3 
# self-upgrade      no
# system            arch=x86_64 os=linux os-distribution=debian os-version=10
# solver            aspcud
# install-criteria  -count(removed),-sum(request,version-lag),-count(down),-sum(solution,version-lag),-count(changed)
# upgrade-criteria  -count(down),-count(removed),-sum(solution,version-lag),-count(new)
# jobs              7
# repositories      1 (http) (default repo at 15d59d7e)
# pinned            0
# current-switch    4.08.0+flambda
$ aspcud --version
aspcud version 1.9.4

License: The MIT License <https://opensource.org/licenses/MIT>
```
It would be cool if opam could say what's wrong with request - which versions or at least libraries conflict.

/CC @AltGr

**Comment by AltGr** [comment]
It seems to be a problem with aspcud: what I can guess is that it returned `no solution`, which causes opam to call the built-in SAT solver from the Dose library to get an explanation why. Of course, if for example there was indeed a solution, we can't get an explanation, and that might have been what happened. Can you try with the built-in mccs solver instead (`--solver=builtin-mccs+glpk`) ? It is in general a better idea anyway, and is the new default (you can customise that in `~/.opam/config`)

**Comment by NightBlues** [comment]
Unfortunately, I can't reproduce the problem because package ppx_tools has now support for 4.08. I also can not use solver you suggest with opam 2.0.3, it says:
```
[ERROR] No installed solver matching the selected 'builtin-mccs+glpk' found
'opam install -vvv --solver=builtin-mccs+glpk ppx_deriving' failed.
```
I've upgraded opam to 2.0.5, and now I have builtin-mccs+glpk solver, but I don't know how to reproduce a problem:
```
$ opam install ppx_tools.5.1+4.06.0
The following dependencies couldn't be met:
  - ppx_tools → ocaml < 4.08
      base of this switch (use `--unlock-base' to force)

No solution found, exiting
$ opam install --solver=aspcud ppx_tools.5.1+4.06.0
The following dependencies couldn't be met:
  - ppx_tools → ocaml < 4.08
      base of this switch (use `--unlock-base' to force)

No solution found, exiting
```

**Comment by NightBlues** [comment]
@AltGr I can reproduce the problem:)
```
~[0]$ opam pin websocket -k git https://github.com/vbmithr/ocaml-websocket.git
[websocket.2.13] synchronised from git+https://github.com/vbmithr/ocaml-websocket.git
websocket is now pinned to git+https://github.com/vbmithr/ocaml-websocket.git (version 2.13)

Sorry, no solution found: there seems to be a problem with your request.

[NOTE] Pinning command successful, but your installed packages may be out of sync.
~[20]$ opam install websocket

<><> Synchronising pinned packages ><><><><><><><><><><><><><><><><><><><><><><>
[websocket.2.13] no changes from git+https://github.com/vbmithr/ocaml-websocket.git

Sorry, no solution found: there seems to be a problem with your request.

No solution found, exiting
~[20]$ opam config report
# opam config report
# opam-version      2.0.5 
# self-upgrade      no
# system            arch=x86_64 os=linux os-distribution=debian os-version=testing
# solver            builtin-mccs+glpk
# install-criteria  -removed,-count[version-lag,request],-count[version-lag,changed],-changed
# upgrade-criteria  -removed,-count[version-lag,solution],-new
# jobs              7
# repositories      1 (http), 1 (version-controlled) (default repo at 631640b4)
# pinned            1 (git)
# current-switch    4.05.0-clean
```
It fails because version 2.13 from git has dependency ocmal >= 4.06 while I have 4.05. But opam doesn't show me this detail, I still have to investigate it manually.

**Comment by AltGr** [comment]
Thanks for the error cases.
The new conflict message engine (#4349) gives these results for @NightBlues case:

```
websocket is now pinned to git+https://github.com/vbmithr/ocaml-websocket.git (version 2.14)

[ERROR] Package conflict!
  * No agreement on the version of ocaml:
    - (invariant) → ocaml-base-compiler = 4.05.0 → ocaml = 4.05.0
    - websocket → ocaml >= 4.06.0
    You can temporarily relax the switch invariant with `--update-invariant'
```

which seems fairly good. I couldn't reproduce @talex5's case though.

**Comment by dra27** [comment]
It looks like this has been fixed in opam 2.1.0 - please do re-open/open-new if qubes-mirage-firewall issue remerges, @talex5!
